File: aeron-client/src/test/java/uk/co/real_logic/aeron/logbuffer/TermRebuilderTest.java
/*
 * Copyright 2014 - 2015 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.aeron.logbuffer;

import org.junit.Before;
import org.junit.Test;
import org.mockito.InOrder;
import uk.co.real_logic.agrona.BitUtil;
import uk.co.real_logic.agrona.concurrent.UnsafeBuffer;

import java.nio.ByteBuffer;

import static java.lang.Integer.valueOf;
import static java.nio.ByteOrder.LITTLE_ENDIAN;
import static org.mockito.Mockito.*;
import static uk.co.real_logic.aeron.logbuffer.FrameDescriptor.*;

public class TermRebuilderTest
{
    private static final int TERM_BUFFER_CAPACITY = LogBufferDescriptor.TERM_MIN_LENGTH;

    private final UnsafeBuffer termBuffer = mock(UnsafeBuffer.class);

    @Before
    public void setUp()
    {
        when(valueOf(termBuffer.capacity())).thenReturn(valueOf(TERM_BUFFER_CAPACITY));
    }

    @Test
    public void shouldInsertIntoEmptyBuffer()
    {
        final UnsafeBuffer packet = new UnsafeBuffer(ByteBuffer.allocateDirect(256));
        final int termOffset = 0;
        final int srcOffset = 0;
        final int length = 256;
        packet.putInt(srcOffset, length, LITTLE_ENDIAN);
        when(termBuffer.getInt(0, LITTLE_ENDIAN)).thenReturn(length);

        TermRebuilder.insert(termBuffer, termOffset, packet, length);

        final InOrder inOrder = inOrder(termBuffer);
        inOrder.verify(termBuffer).putBytes(termOffset, packet, srcOffset, length);
        inOrder.verify(termBuffer).putIntOrdered(termOffset, length);
    }

    @Test
    public void shouldInsertLastFrameIntoBuffer()
    {
        final int frameLength = BitUtil.align(256, FRAME_ALIGNMENT);
        final int srcOffset = 0;
        final int tail = TERM_BUFFER_CAPACITY - frameLength;
        final int termOffset = tail;
        final UnsafeBuffer packet = new UnsafeBuffer(ByteBuffer.allocateDirect(frameLength));
        packet.putShort(typeOffset(srcOffset), (short)PADDING_FRAME_TYPE, LITTLE_ENDIAN);
        packet.putInt(srcOffset, frameLength, LITTLE_ENDIAN);

        when(termBuffer.getInt(tail, LITTLE_ENDIAN)).thenReturn(frameLength);
        when(termBuffer.getShort(typeOffset(tail), LITTLE_ENDIAN)).thenReturn((short)PADDING_FRAME_TYPE);

        TermRebuilder.insert(termBuffer, termOffset, packet, frameLength);

        verify(termBuffer).putBytes(tail, packet, srcOffset, frameLength);
    }

    @Test
    public void shouldFillSingleGap()
    {
        final int frameLength = 50;
        final int alignedFrameLength = BitUtil.align(frameLength, FRAME_ALIGNMENT);
        final int srcOffset = 0;
        final int tail = alignedFrameLength;
        final int termOffset = tail;
        final UnsafeBuffer packet = new UnsafeBuffer(ByteBuffer.allocateDirect(alignedFrameLength));

        when(termBuffer.getInt(0)).thenReturn(frameLength);
        when(termBuffer.getInt(alignedFrameLength, LITTLE_ENDIAN)).thenReturn(frameLength);
        when(termBuffer.getInt(alignedFrameLength * 2, LITTLE_ENDIAN)).thenReturn(frameLength);

        TermRebuilder.insert(termBuffer, termOffset, packet, alignedFrameLength);

        verify(termBuffer).putBytes(tail, packet, srcOffset, alignedFrameLength);
    }

    @Test
    public void shouldFillAfterAGap()
    {
        final int frameLength = 50;
        final int alignedFrameLength = BitUtil.align(frameLength, FRAME_ALIGNMENT);
        final int srcOffset = 0;
        final UnsafeBuffer packet = new UnsafeBuffer(ByteBuffer.allocateDirect(alignedFrameLength));
        final int termOffset = alignedFrameLength * 2;

        when(termBuffer.getInt(0, LITTLE_ENDIAN)).thenReturn(0);
        when(termBuffer.getInt(alignedFrameLength, LITTLE_ENDIAN)).thenReturn(frameLength);

        TermRebuilder.insert(termBuffer, termOffset, packet, alignedFrameLength);

        verify(termBuffer).putBytes(alignedFrameLength * 2, packet, srcOffset, alignedFrameLength);
    }

    @Test
    public void shouldFillGapButNotMoveTailOrHwm()
    {
        final int frameLength = 50;
        final int alignedFrameLength = BitUtil.align(frameLength, FRAME_ALIGNMENT);
        final int srcOffset = 0;
        final UnsafeBuffer packet = new UnsafeBuffer(ByteBuffer.allocateDirect(alignedFrameLength));
        final int termOffset = alignedFrameLength * 2;

        when(termBuffer.getInt(0, LITTLE_ENDIAN)).thenReturn(frameLength);
        when(termBuffer.getInt(alignedFrameLength, LITTLE_ENDIAN)).thenReturn(0);

        TermRebuilder.insert(termBuffer, termOffset, packet, alignedFrameLength);

        verify(termBuffer).putBytes(alignedFrameLength * 2, packet, srcOffset, alignedFrameLength);
    }
}

File: aeron-client/src/test/cpp/concurrent/TermRebuilderTest.cpp
/*
 * Copyright 2014-2018 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


#include <gtest/gtest.h>

#include <thread>
#include "MockAtomicBuffer.h"
#include <concurrent/logbuffer/DataFrameHeader.h>
#include <concurrent/logbuffer/TermRebuilder.h>
#include <concurrent/logbuffer/LogBufferDescriptor.h>

using namespace aeron::concurrent::logbuffer;
using namespace aeron::concurrent::mock;
using namespace aeron::concurrent;
using namespace aeron;

#define TERM_BUFFER_CAPACITY (LogBufferDescriptor::TERM_MIN_LENGTH)

typedef std::array<std::uint8_t, TERM_BUFFER_CAPACITY> log_buffer_t;
typedef std::array<std::uint8_t, 256> packet_t;

class TermRebuilderTest : public testing::Test
{
public:
    TermRebuilderTest() :
        m_log(&m_logBuffer[0], m_logBuffer.size()), m_packet(m_hdrBuffer, 0)
    {
        m_logBuffer.fill(0);
    }

    virtual void SetUp()
    {
        m_logBuffer.fill(0);
    }

protected:
    AERON_DECL_ALIGNED(log_buffer_t m_logBuffer, 16);
    AERON_DECL_ALIGNED(packet_t m_hdrBuffer, 16);
    MockAtomicBuffer m_log;
    AtomicBuffer m_packet;
};

TEST_F(TermRebuilderTest, shouldInsertIntoEmptyBuffer)
{
    const util::index_t termOffset = 0;
    const util::index_t srcOffset = 0;
    const util::index_t length = 256;

    m_packet.putInt32(srcOffset, length);

    ON_CALL(m_log, getInt32(0)).WillByDefault(testing::Return(length));

    testing::Sequence sequence;

    EXPECT_CALL(m_log, putBytes(termOffset, testing::Ref(m_packet), srcOffset, length)).InSequence(sequence);
    EXPECT_CALL(m_log, putInt32Ordered(termOffset, length)).InSequence(sequence);

    TermRebuilder::insert(m_log, termOffset, m_packet, length);
}

TEST_F(TermRebuilderTest, shouldInsertLastFrameIntoBuffer)
{
    const std::int32_t frameLength = util::BitUtil::align(256, FrameDescriptor::FRAME_ALIGNMENT);
    const std::int32_t srcOffset = 0;
    const std::int32_t tail = TERM_BUFFER_CAPACITY - frameLength;
    const std::int32_t termOffset = tail;

    m_packet.putUInt16(FrameDescriptor::typeOffset(srcOffset), DataFrameHeader::HDR_TYPE_PAD);
    m_packet.putInt32(srcOffset, frameLength);

    ON_CALL(m_log, getInt32(tail))
        .WillByDefault(testing::Return(frameLength));
    ON_CALL(m_log, getUInt16(FrameDescriptor::typeOffset(tail)))
        .WillByDefault(testing::Return(DataFrameHeader::HDR_TYPE_PAD));

    EXPECT_CALL(m_log, putBytes(tail, testing::Ref(m_packet), srcOffset, frameLength));
    EXPECT_CALL(m_log, putInt32Ordered(tail, frameLength));

    TermRebuilder::insert(m_log, termOffset, m_packet, frameLength);
}

TEST_F(TermRebuilderTest, shouldFillSingleGap)
{
    const std::int32_t frameLength = 50;
    const std::int32_t alignedFrameLength = util::BitUtil::align(frameLength, FrameDescriptor::FRAME_ALIGNMENT);
    const std::int32_t srcOffset = 0;
    const std::int32_t tail = alignedFrameLength;
    const std::int32_t termOffset = tail;

    ON_CALL(m_log, getInt32(alignedFrameLength)).WillByDefault(testing::Return(frameLength));
    ON_CALL(m_log, getInt32(alignedFrameLength * 2)).WillByDefault(testing::Return(frameLength));

    EXPECT_CALL(m_log, putBytes(tail, testing::Ref(m_packet), srcOffset, alignedFrameLength));
    EXPECT_CALL(m_log, putInt32Ordered(alignedFrameLength, srcOffset));

    TermRebuilder::insert(m_log, termOffset, m_packet, alignedFrameLength);
}

TEST_F(TermRebuilderTest, shouldFillAfterGap)
{
    const std::int32_t frameLength = 50;
    const std::int32_t alignedFrameLength = util::BitUtil::align(frameLength, FrameDescriptor::FRAME_ALIGNMENT);
    const std::int32_t srcOffset = 0;
    const std::int32_t termOffset = alignedFrameLength * 2;

    ON_CALL(m_log, getInt32(0)).WillByDefault(testing::Return(0));
    ON_CALL(m_log, getInt32(alignedFrameLength)).WillByDefault(testing::Return(frameLength));

    EXPECT_CALL(m_log, putBytes(alignedFrameLength * 2, testing::Ref(m_packet), srcOffset, alignedFrameLength));
    EXPECT_CALL(m_log, putInt32Ordered(alignedFrameLength * 2, 0));

    TermRebuilder::insert(m_log, termOffset, m_packet, alignedFrameLength);
}

TEST_F(TermRebuilderTest, shouldFillGapButNotMoveTailOrHwm)
{
    const std::int32_t frameLength = 50;
    const std::int32_t alignedFrameLength = util::BitUtil::align(frameLength, FrameDescriptor::FRAME_ALIGNMENT);
    const std::int32_t srcOffset = 0;
    const std::int32_t termOffset = alignedFrameLength * 2;

    ON_CALL(m_log, getInt32(0)).WillByDefault(testing::Return(frameLength));
    ON_CALL(m_log, getInt32(alignedFrameLength)).WillByDefault(testing::Return(0));

    EXPECT_CALL(m_log, putBytes(alignedFrameLength * 2, testing::Ref(m_packet), srcOffset, alignedFrameLength));
    EXPECT_CALL(m_log, putInt32Ordered(alignedFrameLength * 2, 0));

    TermRebuilder::insert(m_log, termOffset, m_packet, alignedFrameLength);
}
